File: rozmovlyalka-js/src/wav.js
// WAV — упаковка PCM в WAV (11025 Гц, моно, 8 бит без знака) с оригинальным
// 44-байтным заголовком программы (wavhead.bin): поля размеров патчатся под
// длину данных.
import { fromHex } from './util.js';

// Точные байты data/wavhead.bin (шаблон DAT_02365704).
const WAV_TEMPLATE = fromHex(
  '524946462400000057415645666d74201000000001000100112b0000112b0000010008006461746100000000',
);

export function wavHeader(dataLen) {
  const h = WAV_TEMPLATE.slice();
  const dv = new DataView(h.buffer);
  dv.setUint32(4, 44 - 8 + dataLen, true); // filesize-8
  dv.setUint32(0x28, dataLen, true); // filesize-44
  return h;
}

// pcmToWAV — PCM → готовый WAV-файл (Uint8Array).
export function pcmToWAV(pcm) {
  const h = wavHeader(pcm.length);
  const out = new Uint8Array(44 + pcm.length);
  out.set(h, 0);
  out.set(pcm, 44);
  return out;
}

File: wav.go
package rozmovlialka

import (
	"encoding/binary"
	"fmt"
	"io"
)

// wavHeader — 44-байтный RIFF-шаблон из DAT_02365704 (wavhead.bin), поля
// размеров (@4 и @0x28) патчатся под длину данных.
func wavHeader(dataLen int, patched [44]byte) []byte {
	h := patched
	binary.LittleEndian.PutUint32(h[4:], uint32(44-8+dataLen)) // filesize-8
	binary.LittleEndian.PutUint32(h[0x28:], uint32(dataLen))   // filesize-44
	return h[:]
}

// loadWavTemplate возвращает копию шаблона wavhead.bin из встроенных данных.
func loadWavTemplate() ([44]byte, error) {
	var t [44]byte
	b, err := gunzipData("wavhead.bin.gz")
	if err != nil {
		return t, err
	}
	if len(b) != 44 {
		return t, fmt.Errorf("rozmovlialka: шаблон wavhead имеет размер %d, ожидалось 44", len(b))
	}
	copy(t[:], b)
	return t, nil
}

// WriteWAV — упаковка PCM в WAV (11025 Гц, моно, 8 бит без знака) с
// оригинальным 44-байтным заголовком программы.
func WriteWAV(w io.Writer, pcm []byte) error {
	t, err := loadWavTemplate()
	if err != nil {
		return err
	}
	if _, err := w.Write(wavHeader(len(pcm), t)); err != nil {
		return err
	}
	_, err = w.Write(pcm)
	return err
}
